### PR TITLE
Nestjs

### PR DESCRIPTION
**Pull request by sophieheb**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ before_script:
     - psql -c 'create database littlewindowtest;' -U postgres
 script:
   - npm run lint
-  # TODO: Add in tests!
-  # - npm test --passWithNoTests
-  # - cd client && npm test --passWithNoTests
+  - npm run testtravis
+  - cd client && npm install
+  - npm run test --passWithNoTests
```
Comment (jits) [inline .travis.yml:14]: It looks like `--passWithNoTests` is [already set in the `package.json`](https://github.com/chaynHQ/little-window/pull/250/files#diff-52dbbf94d5c54b68c0cad0702ba94a96R22) so probably don't need it here.

File: client/package.json [hunk 2]
```diff
@@ -16,10 +17,11 @@
...
   "scripts": {
-    "start": "react-scripts start",
-    "build": "react-scripts build",
-    "test": "react-scripts test --env=jsdom",
-    "eject": "react-scripts eject"
+    "start": "node_modules/.bin/react-scripts start",
+    "build": "node_modules/.bin/react-scripts build",
+    "test": "node_modules/.bin/react-scripts test --env=jsdom --passWithNoTests",
+    "eject": "node_modules/.bin/react-scripts eject",
+    "lint": "eslint '**/*.js' --fix "
```
Comment (jits) [inline client/package.json:24]: I'd suggest splitting out the `lint`ing and fixing into two separate tasks – `lint` and `lint-fix`. The CI should only ever run the `lint` task (and not modify anything).

File: package.json [hunk 3]
```diff
@@ -2,47 +2,106 @@
   "name": "little-window",
   "version": "1.0.0",
   "description": "A chatbot featured on Chayn websites, which quickly directs users to Chayn’s relevant internal or external resources",
-  "main": "config.js",
+  "author": "Chayn",
+  "license": "ISC",
```
Comment (jits) [inline package.json:6]: (I'm guessing this was autogenerated / updated by the NestJS CLI) The top level license is MIT so we should change this.

```suggestion
  "license": "MIT",
```

File: src/app.module.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  providers: [],
+  imports: [
+    ConfigModule.forRoot({ isGlobal: true }),
+    TypeOrmModule.forRoot(databaseConfig),
+    ConversationModule,
+    UserMessageModule,
+    MessageModule,
+    LoggerModule.forRoot({
+      accessToken: process.env.ROLLBAR_TOKEN,
+      environment: process.env.NODE_ENV,
```
Comment (jits) [inline src/app.module.ts:25]: This should probably be a separate ENV var that provides the _application environment_ as opposed to the Node environment? (i.e. should be `preview` or `staging` or `production`, or `review-12345` if a review app).

File: package.json [hunk 5]
```diff
@@ -2,47 +2,94 @@
...
-    "start": "node ./src/server.js",
-    "start:test": "NODE_ENV=production node ./src/server.js",
-    "postinstall": "cd client && npm install && npm run build",
-    "test": "NODE_ENV=test tape ./src/test/*test.js | tap-spec && cd client && npm run test",
-    "server": "NODE_ENV=dev nodemon ./src/server.js",
-    "dev": "concurrently \"npm run server\" \"npm run client\"",
-    "client": "cd client && npm start",
-    "lint": "eslint '**/*.js'"
+    "prebuild": "rimraf dist",
+    "postinstall": "nest build",
```
Comment (jits) [inline package.json:9]: It's worth still having `"build": "nest build"` as a separate task, and then this postinstall can just do `npm build`. (Just so it's clear it's not Heroku specific).

File: package.json [hunk 6]
```diff
@@ -2,47 +2,94 @@
...
+    "heroku-postbuild": "cd client && npm install && npm run build",
```
Comment (jits) [inline package.json:10]: Similar to my above comment, it's worth moving these commands into a `build-client` tasks, which can then be called here via `npm build-client` (also makes it clear what exactly is going on).

File: client/package.json (hunk 2 above)
Comment (sophieheb) [reply client/package.json]: Update made

File: .travis.yml (hunk 1 above)
Comment (sophieheb) [reply .travis.yml]: Update Made

File: src/app.module.ts (hunk 4 above)
Comment (sophieheb) [reply src/app.module.ts:27]: Updated

File: package.json (hunk 5 above)
Comment (sophieheb) [reply package.json:9]: Updated!

File: src/botMessage/dialogFlow.service.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    userMessage: string,
+  ): Promise<string> {
+    let privateKey;
+    if (process.env.NODE_ENV === 'production'){
+      privateKey = JSON.parse(process.env.DIALOGFLOW_PRIVATE_KEY)
+    } else if (process.env.NODE_ENV === 'travis') {
+      privateKey = process.env.DIALOGFLOW_PRIVATE_KEY.replace(/\\n/gm, '\n')
+    } else {
+      process.env.DIALOGFLOW_PRIVATE_KEY
+    }
```
Comment (jits) [inline src/botMessage/dialogFlow.service.ts:17]: This processing will happen every time this method is called – which may be fine, if it's only a few times. But if it's a regular method call, I'd suggest pulling this out to the top so it only happens once, on loading this file.

(Also note, for the `else` branch it doesn't look like `privateKey` is being set?)

File: src/botMessage/dialogFlow.service.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        'client_email': clientEmail,
+      },
+    };
+    const sessionClient = new dialogflow.SessionsClient(config);
+    const sessionPath = sessionClient.sessionPath(
+      process.env.DIALOGFLOW_PROJECT_ID,
+      conversationId,
+    );
+
+    // TODO: set language Code
```
Comment (jits) [inline src/botMessage/dialogFlow.service.ts:32]: Looks like this `TODO` could be removed now?

File: src/conversation/conversation.service.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      this.rollbarLogger.error(
+        {
+          message: 'Cannot find a conversation with id: ' + conversationId,
+          name: 'No ConversationId',
+        },
+        'Get Conversation',
+      );
+    }
+
+    return column ? conversation[column] : conversation;
```
Comment (jits) [inline src/conversation/conversation.service.ts:52]: If `conversation` was not found, this line will throw an error, or return `null`. We may want to trigger a `404 Not Found` in the controller.

On a more general note, is there an error handling pattern/mechanic in NestJS to automatically handle not found entities from the db? (E.g. in Rails there's the ability for the entity find to throw a specific exception that gets caught higher up the stack and gets handled [and logged/reported] and a `404` response is rendered).

File: src/userMessage/userMessage.service.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    // so we don't have to filter afterwards.
+    const splitUserResponse = userResponse.split('-');
+    const botResponses = await this.storyblokService.getBotResponsesBySlug(
+      'setup',
+    );
+
+    const previousMessageWasSetupMessage =
+      botResponses.filter(
+        response => response['uuid'] === previousMessageStoryblokId,
+      ).length > 0;
```
Comment (jits) [inline src/userMessage/userMessage.service.ts:27]: An alternative (achieves the same thing):

```suggestion
    const previousMessageWasSetupMessage =
      botResponses.some(
        response => response['uuid'] === previousMessageStoryblokId,
      );
```

File: src/botMessage/dialogFlow.service.ts (hunk 7 above)
Comment (sophieheb) [reply src/botMessage/dialogFlow.service.ts]: Updated.

File: src/botMessage/dialogFlow.service.ts (hunk 8 above)
Comment (sophieheb) [reply src/botMessage/dialogFlow.service.ts:33]: Not yet! I need to do some French updates then we'll have to switch dialogflow back and forth.

File: src/conversation/conversation.service.ts (hunk 9 above)
Comment (sophieheb) [reply src/conversation/conversation.service.ts:52]: NestJs automatically handles errors as they float up the chain to the response so a 404 will automatically happen in this case, but it should be caught by the rollbar error above also.

File: src/botMessage/dialogFlow.service.ts (hunk 8 above)
Comment (jits) [reply src/botMessage/dialogFlow.service.ts:33]: Ah! Good to know. The comment could do with a little more elaboration (based on what you said). I know I would forgot in a few weeks times if/when I see this comment 😄

**Pull request merged**